### PR TITLE
amrex: support sundials variant in newer amrex versions

### PR DESCRIPTION
**Pull request by balos1**
AMReX 21.07 reintroduces sundials interfaces, albeit entirely different ones. 

CC: @balay

### CODE REVIEW

**Comment by balos1** [comment]
@spackbot maintainers

**Comment by WeiqunZhang** [comment]
I failed to build with `spack install amrex +sundials +cuda`.  It died when building sundials.  The build log is attached.  Note that `spack install amrex +cuda` works for me.

[spack-build-out.txt](https://github.com/spack/spack/files/7102103/spack-build-out.txt)

**Comment by balos1** [comment]
@WeiqunZhang The problem is that sundials does not like `cuda_arch=none` (which is the default value). `spack install amrex +cuda +sundials ^sundials cuda_arch=xx` works for me.  I will add a conflict in the sundials package for the case that `cuda_arch=none`. 

Now the question is, should we pass the AMReX `cuda_arch` down to sundials like this:

```
diff --git a/var/spack/repos/builtin/packages/amrex/package.py b/var/spack/repos/builtin/packages/amrex/package.py
index 6e20867cdb..7ead93ca56 100644
--- a/var/spack/repos/builtin/packages/amrex/package.py
+++ b/var/spack/repos/builtin/packages/amrex/package.py
@@ -83,7 +83,8 @@ class Amrex(CMakePackage, CudaPackage, ROCmPackage):
     depends_on('mpi', when='+mpi')
     depends_on('sundials@4.0.0:4.1.0 +ARKODE +CVODE', when='@19.08:20.11 +sundials')
     depends_on('sundials@5.7.0: +ARKODE +CVODE', when='@21.07: +sundials')
-    depends_on('sundials@5.7.0: +ARKODE +CVODE +cuda', when='@21.07: +sundials +cuda')
+    for arch in CudaPackage.cuda_arch_values:
+        depends_on(f'sundials@5.7.0: +ARKODE +CVODE +cuda cuda_arch={arch}', when=f'@21.07: +sundials +cuda cuda_arch={arch}')
     depends_on('cuda@9.0.0:', when='+cuda')
     depends_on('python@2.7:', type='build', when='@:20.04')
     depends_on('cmake@3.5:',  type='build', when='@:18.10.99')
```

Or, should we just require the `spack install amrex +cuda +sundials ^sundials cuda_arch=xx` syntax?

File: var/spack/repos/builtin/packages/amrex/package.py [hunk 1]
```diff
@@ -68,20 +68,22 @@ class Amrex(CMakePackage, CudaPackage, ROCmPackage):
...
+            description='Enable SUNDIALS interfaces')
     variant('pic', default=False,
             description='Enable PIC')
 
     # Build dependencies
     depends_on('mpi', when='+mpi')
-    depends_on('sundials@4.0.0:4.1.0 +ARKODE +CVODE', when='@19.08: +sundials')
+    depends_on('sundials@4.0.0:4.1.0 +ARKODE +CVODE', when='@19.08:20.11 +sundials')
+    depends_on('sundials@5.7.0: +ARKODE +CVODE', when='@21.07: +sundials')
+    depends_on('sundials@5.7.0: +ARKODE +CVODE +cuda', when='@21.07: +sundials +cuda')
```
Comment (sayerhs) [inline var/spack/repos/builtin/packages/amrex/package.py]:
```suggestion
    for arch in CudaPackage.cuda_arch_values:
        depends_on('sundials@5.7.0: +ARKODE +CVODE +cuda cuda_arch=%s' % arch, when='@21.07: +sundials +cuda cuda_arch=%s' % arch)
```

Comment (WeiqunZhang) [reply var/spack/repos/builtin/packages/amrex/package.py]: @sayerhs's suggestion works.  I can install with `spack install amrex +cuda +sundials cuda_arch=xx` without extra `^sundials`.

Comment (WeiqunZhang) [reply var/spack/repos/builtin/packages/amrex/package.py]: @balos1  Let's pass the amerx cuda_arch to sundials.

Comment (balos1) [reply var/spack/repos/builtin/packages/amrex/package.py]: @WeiqunZhang done.

File: var/spack/repos/builtin/packages/amrex/package.py [hunk 2]
```diff
@@ -68,20 +68,23 @@ class Amrex(CMakePackage, CudaPackage, ROCmPackage):
...
     variant('pic', default=False,
             description='Enable PIC')
 
     # Build dependencies
     depends_on('mpi', when='+mpi')
-    depends_on('sundials@4.0.0:4.1.0 +ARKODE +CVODE', when='@19.08: +sundials')
+    depends_on('sundials@4.0.0:4.1.0 +ARKODE +CVODE', when='@19.08:20.11 +sundials')
+    depends_on('sundials@5.7.0: +ARKODE +CVODE', when='@21.07: +sundials')
+    for arch in CudaPackage.cuda_arch_values:
+        depends_on(f'sundials@5.7.0: +ARKODE +CVODE +cuda cuda_arch={arch}', when=f'@21.07: +sundials +cuda cuda_arch={arch}')
```
Comment (sayerhs) [inline var/spack/repos/builtin/packages/amrex/package.py]: You might want to check with spack devs if they are still supporting Python 3.5. if I remember correctly, f-strings only work on Python>=3.6

Comment (balos1) [reply var/spack/repos/builtin/packages/amrex/package.py]: I see 3.5 in the unit tests as well as 2.7, so I will switch it.

File: var/spack/repos/builtin/packages/amrex/package.py [hunk 3]
```diff
@@ -68,20 +68,23 @@ class Amrex(CMakePackage, CudaPackage, ROCmPackage):
...
     variant('pic', default=False,
             description='Enable PIC')
 
     # Build dependencies
     depends_on('mpi', when='+mpi')
-    depends_on('sundials@4.0.0:4.1.0 +ARKODE +CVODE', when='@19.08: +sundials')
+    depends_on('sundials@4.0.0:4.1.0 +ARKODE +CVODE', when='@19.08:20.11 +sundials')
+    depends_on('sundials@5.7.0: +ARKODE +CVODE', when='@21.07: +sundials')
+    for arch in CudaPackage.cuda_arch_values:
+        depends_on('sundials@5.7.0: +ARKODE +CVODE +cuda cuda_arch=%s' % arch, when='@21.07: +sundials +cuda cuda_arch=%s' % arch)
```
Comment (sayerhs) [inline var/spack/repos/builtin/packages/amrex/package.py:87]: @balos1 What is the expected outcome when doing `spack install amrex~mpi`? I expect it should link to `sundials~mpi` correct? We might want to add extra depend rules (something like below)

https://github.com/spack/spack/blob/c6c9213766c5755ae893d202d1336883ad2296fa/var/spack/repos/builtin/packages/amr-wind/package.py#L68-L72

Comment (balos1) [reply var/spack/repos/builtin/packages/amrex/package.py:87]: Currently, the AMReX interfaces to SUNDIALS do not involve MPI, so any combination of `amrex+mpi` and `amrex~mpi` with `sundials+mpi` or `sundials~mpi` should be fine (I just tested them to confirm). So I think we should leave it unspecified.

Comment (balos1) [reply var/spack/repos/builtin/packages/amrex/package.py:87]: We do want to add similar support for `+rocm` like `+cuda`.

Comment (WeiqunZhang) [reply var/spack/repos/builtin/packages/amrex/package.py:87]: That would be great!  Do you want to do it in this PR or a follow-up one?

Comment (balos1) [reply var/spack/repos/builtin/packages/amrex/package.py:87]: I think we should try and include it in this PR. I am attempting to build amrex +rocm on Spock (so far unsuccessful), will follow up with more details on slack.

Comment (balos1) [reply var/spack/repos/builtin/packages/amrex/package.py:87]: OK, I pushed up support for `+rocm`  and verified it builds on spock. I think this PR should be good to go now.

**Review by WeiqunZhang — APPROVED** [review]

**Review by ax3l — APPROVED** [review]
Thanks a lot! :tada:

**Pull request merged**